[git-webkit] Speed up Branch.branch_point

JonWBedard opened the pull request:
#### 295115e98b2ed780c75595560254ee048755717a
<pre>
[git-webkit] Speed up Branch.branch_point
<a href="https://bugs.webkit.org/show_bug.cgi?id=256855">https://bugs.webkit.org/show_bug.cgi?id=256855</a>
rdar://109419012

Reviewed by Elliott Williams.

Calculating the merge-base of a commit and all production branches is actually faster than checking
which branches that commit is a member of. Re-implement branch-point using this techinque.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/git-repo.json: Add identifiers to all commits.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Pass all arguments to merge-base
(Git.merge_base): Support multiple branch arguments.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.branch_point): Calculate the merge-base between the current commit and all potential production branches.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.pull_request_branch_point): Early-exit function if branch-point is undefined.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:
(TestBranch.test_branch_point):
(TestBranch.test_branch_point_main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/squash_unittest.py:

Canonical link: <a href="https://commits.webkit.org/264536@main">https://commits.webkit.org/264536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a5c161eec0d824e54be8e80b62e8a6f3c73a590

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/7942 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/8221 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/8408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/9590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/7946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/10212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/8136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/9590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/8083 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/10212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/8408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/9709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/8068 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/10212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/8408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/9709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/10212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/8408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/9709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/7859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/8136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/7903 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/7176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/8408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/11386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/947 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/7595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -83,27 +87,67 @@ def editable(cls, branch, repository=None):
+                    cwd=repository.root_path,
+                    capture_output=True,
+                ).returncode:
+                    break
```
I'm not particularly keen on calling this repeatedly in the hot loop? `merge-base` in any form is potentially expensive (as worst-case has to walk the entire parent DAG from `repository.default_branch` for each `commit_hash`).

I'd rather move `git merge-base HEAD {repository.default_branch}` to being outside the loop, and then we can simply check if we've reached the merge base yet within the loop.

We can also then limit the scope of the `rev-list` call to the commits between the merge base and HEAD.

JonWBedard replied:
Oh, good point, yeah, we can avoid this call by doing that.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -83,27 +87,67 @@ def editable(cls, branch, repository=None):
+                    cwd=repository.root_path,
+                    capture_output=True,
+                    encoding='utf-8',
+                ).stdout.splitlines():
```
Given how expensive this is (taking many seconds per call, as it has to walk the entire DAG for each call), could we instead look at the commit message and pull out the branch from the canonical link? Because that branch is likely to contain this commit and is likely to not be editable.

We'll still need to fall back to this, but in the common case we can avoid the expensive call.

JonWBedard replied:
I think falling back isn't going to help, because you've got the common case backwards. The common case is actually that the commit we're referencing isn't on a production branch, it's on a PR branch. In that case (as you point out) we have to fall back to the more expensive call anyways.

It's also worth pointing out that this call is as expensive as the number of branches which contain a commit. That will be most dramatic if the commit is on `main`. Except we won't run this code if the commit is on `main` (because the `merge-base` call will have caught that case).

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -83,27 +87,67 @@ def editable(cls, branch, repository=None):
+                stderr=subprocess.PIPE,
+                cwd=repository.root_path,
+                **(dict(encoding='utf-8') if sys.version_info >= (3, 6) else dict())
+            )
```
Can we add a comment making it clear why we're streaming this? (Presumably performance, given we're requesting the entire rev-list of HEAD.)

gsnedders reviewed (commented):
I wonder if we can go even more radical than this change?

Do we have a way to get a list of non-editable branches reasonably quickly? Because if we do, we could do something like `git merge-base HEAD {non-editable branches…}` and avoid invoking multiple git commands all of which might walk the DAG themselves?

I feel like I proposed this somewhere before, but I can't find any reference to it right now.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -90,20 +91,57 @@ def editable(cls, branch, repository=None):
-            else:
-                log.info('    No commits on editable branch')
+    def branch_point(cls, repository):
+        branches = repository.branches_for(remote=None)
+        production_branches = [
+            'remotes/{}/{}'.format(remote, branch)
+            for remote in repository.source_remotes()
+            for branch in branches[remote] if not repository.dev_branches.match(branch)
+        ]
```
This feels like something we could just put on `Repository` rather than having it inline here?

JonWBedard replied:
This is the only place we need this particular list, at the moment. I'm not against moving this to `Repository` in the future, but for the moment, I don't. see where else we'd need to use this.

gsnedders replied:
Yeah, no strong opinion here.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -90,20 +91,57 @@ def editable(cls, branch, repository=None):
+        ).stdout.strip()
+
+        partial_bases = set()
+        for shard in [
+            production_branches[cls.MERGE_BASE_SHARD_SIZE * i:cls.MERGE_BASE_SHARD_SIZE * (i + 1)]
+            for i in range(1 + len(production_branches) // cls.MERGE_BASE_SHARD_SIZE)
+        ]:
```
Do we not have something like the [`itertools` `grouper` example](https://docs.python.org/3/library/itertools.html#itertools-recipes) anywhere?

JonWBedard replied:
Not that I'm aware of...I'm open to simpler ways to do this, but this seemed relatively straight forward.

gsnedders replied:
Oh, yeah, this is relatively straight forward, I'm just slightly surprised we don't have something generic to make this slightly simpler in-line.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -90,20 +91,57 @@ def editable(cls, branch, repository=None):
+                encoding='utf-8',
+            )
+            if result.returncode:
+                continue
```
Why do we want to ignore all cases where this exits uncleanly? At the very least adding a comment here would be good.

JonWBedard replied:
It's more of a "what else can we do?" situation. Callers of this function treat "None" as an error, so if we want to represent an error, that's what we should return. Which we could right here. But in a hypothetical situation where one shard fails but the others don't (and I'm not sure how you would even get into such a state), it's a good bet that one of the successful shards actually has the answer you're looking for.

Put another way: if all call to `merge-base` fail, we return `None` (which is the error state). If only one fails, what's the correct action? This code attempts to continue if a single shard fails.

gsnedders replied:
I think I'd rather just return `None` if any fail, because as you say it shouldn't happen, and from my point-of-view it's unclear what the result should be.

JonWBedard replied:
I'm fine with that behavior, will modify

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -90,20 +91,64 @@ def editable(cls, branch, repository=None):
 
-        return commit
+        partial_bases = set()
+        for shard in [
+            production_branches[cls.MERGE_BASE_SHARD_SIZE * i:cls.MERGE_BASE_SHARD_SIZE * (i + 1)]
+            for i in range(1 + len(production_branches) // cls.MERGE_BASE_SHARD_SIZE)
+        ]:
+            if not shard:
+                continue
+            result = run(
+                [repository.executable(), 'merge-base', head] + shard,
+                cwd=repository.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            )
```
I wonder if you could use xargs? Git for Windows [appears to have it](https://github.com/git-for-windows/busybox-w32/blob/main/findutils/xargs.c), so while I don't know how to invoke it from Windows, I'd think we have it available on all ports.

If so, no need for sharding logic, just run

    xargs -0 git merge-base {head}

and pipe in `'\0'.join(production_branches)`.

JonWBedard replied:
I don't think `xargs` is going to help us much here. That command would end up running `git merge-base` on every ref in our list. We would still have to merge the results in a single `git merge-base` command, and we'd still have to be prepared to shard that command because it's possible that even after de-duping, we would have more than 500 merge bases. There is also the issue that running `git merge-base` 1000 times with 1000 arguments is going to be slower than running it 3 times, even if 2 of those times have 500 arguments.

emw-apple replied:
> I don't think `xargs` is going to help us much here. That command would end up running `git merge-base` on every ref in our list

I don't think so – xargs runs as few invocations as possible, respecting ARG_MAX. From xargs(1):

	     The xargs utility reads space, tab, newline and end-of-file delimited
	     strings from the standard input and executes utility with the strings as
	     arguments.
	
	     Any arguments specified on the command line are given to utility upon
	     each invocation, followed by some number of the arguments read from the
	     standard input of xargs.  This is repeated until standard input is
	     exhausted.

iiuc it does exactly what we're doing, except by counting bytes and comparing to the platform's ARG_MAX, not sharding over a fixed number of arguments. So it would potentially reduce the number of merge-base invocations needed.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -90,20 +91,64 @@ def editable(cls, branch, repository=None):
+                break
+            partial_base = result.stdout.strip()
+            if partial_base == head:
+                return repository.commit(
+                    hash=head,
+                    include_log=False, include_identifier=False,
+                )
```
A comment on why it's safe to prematurely return here may be helpful. IIUC, if, for instance, HEAD is the same commit as remotes/origin/main, then we already know that HEAD is the branch point and don't need to consider results from other shards.

emw-apple reviewed (approved):
Looks great!

JonWBedard committed:
[git-webkit] Speed up Branch.branch_point
https://bugs.webkit.org/show_bug.cgi?id=256855
rdar://109419012

Reviewed by Elliott Williams.

Calculating the merge-base of a commit and all production branches is actually faster than checking
which branches that commit is a member of. Re-implement branch-point using this techinque.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/git-repo.json: Add identifiers to all commits.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Pass all arguments to merge-base
(Git.merge_base): Support multiple branch arguments.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.branch_point): Calculate the merge-base between the current commit and all potential production branches.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.pull_request_branch_point): Early-exit function if branch-point is undefined.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:
(TestBranch.test_branch_point):
(TestBranch.test_branch_point_main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/squash_unittest.py:

Canonical link: https://commits.webkit.org/264536@main
